src/ast/nodes/Identifier.ts

```diff
@@ -71,11 +79,48 @@ export default class Identifier extends NodeBase implements PatternNode {
 		}
 	}
 
+	private checkVarRedeclarationInScope(scope: Scope, disallowParameter: boolean) {
```

Comment:
This logic looks quite complicated with many special cases. Did you consider moving the check into the scopes themselves in `addDeclaration`? (Maybe after writing more tests).
Scopes also know how hoisting works, which should allow you to distribute these checks nicely over the different types of scopes.